File: Algorithms/C++/Medium/163_Missing_Ranges.cpp
class Solution {
public:
    vector<string> findMissingRanges(vector<int>& nums, int lower, int upper) {
        vector<string> missingRanges;

        if (lower > upper) {
            return missingRanges;
        }

        if (nums.empty()) {
            missingRanges.push_back(getRange(lower, upper));
            return missingRanges;
        }

        int left, right;
        string range;
        for (int i = 0; i < nums.size(); i++) {
            int num = nums[i];

            left = i == 0 ? lower : nums[i - 1] + 1;
            right = num - 1;

            range = getRange(left, right);
            if (!range.empty()) {
                missingRanges.push_back(range);
            }
        }

        if (nums[nums.size() - 1] < upper) {
            left = nums[nums.size() - 1] + 1;
            right = upper;

            range = getRange(left, right);
            if (!range.empty()) {
                missingRanges.push_back(range);
            }
        }

        return missingRanges;
    }
private:
    string getRange(int left, int right) {
        if (left > right) {
            return "";
        } else if (left == right) {
            return to_string(left);
        } else {
            return to_string(left) + "->" + to_string(right);
        }
    }
};

File: Algorithms/Swift/Medium/163_Missing_Ranges.swift
class Solution {
    func findMissingRanges(nums: [Int], _ lower: Int, _ upper: Int) -> [String] {
        var ranges = [String]()
        
        if lower > upper {
            return ranges
        }
        
        if nums.isEmpty {
            return [getRange(lower, upper)!]
        }
        
        var start = lower, end = nums[0] - 1
        if let range = getRange(start, end) {
            ranges.append(range)
        }
        
        for index in 0..<nums.count - 1 {
            start = nums[index] + 1
            end   = nums[index + 1] - 1
            
            if let range = getRange(start, end) {
                ranges.append(range)
            }
        }
        
        start = nums[nums.count - 1] + 1
        end = upper
        if let range = getRange(start, end) {
            ranges.append(range)
        }
        
        return ranges
    }
    
    private func getRange(start: Int, _ end: Int) -> String? {
        if start > end {
            return nil
        } else if start == end {
            return String(start)
        } else {
            return "\(start)->\(end)"
        }
    }
}
